src/tasks/worker.ts

```diff
@@ -0,0 +1,278 @@
+					}
+				});
+			} catch (err) {
+				// Ignore closed/rollback errors
+				if (!(err instanceof TransactionClosedError)) {
+					throw err;
+				}
+			} finally {
+				// Update task with results
+				await this.finalize(tx, task, startedOnTime, status, error);
+			}
+		} catch (err) {
+			// This shouldn't happen, but if it does we want to log and kill the process
+			console.error(
+				`Failed to execute task ${task.id} with handler ${task.is_executed_by__handler}:`,
+				err,
+			);
+			process.exit(1);
+		} finally {
+			this.executing--;
+		}
+	}
+
+	// Update task and schedule next attempt if needed
+	private async finalize(
+		tx: Db.Tx,
+		task: PartialTask,
+		startedOnTime: Date,
+		status: TaskStatus,
+		errorMessage?: string,
+	): Promise<void> {
+		const attemptCount = task.attempt_count + 1;
+		const body: AnyObject = {
+			started_on__time: startedOnTime,
+			ended_on__time: new Date(),
+			status,
+			attempt_count: attemptCount,
+			...(errorMessage != null && { error_message: errorMessage }),
+		};
+
+		// Re-enqueue if the task failed but has retries left, remember that
+		// attemptCount includes the initial attempt while attempt_limit does not
+		if (status === 'failed' && attemptCount < task.attempt_limit) {
+			body.status = 'queued';
+
+			// Schedule next attempt using exponential backoff
+			body.is_scheduled_to_execute_on__time =
+				this.getNextAttemptTime(attemptCount);
+		}
+
+		// Patch current task
+		await this.client.patch({
+			resource: 'task',
+			passthrough: {
+				tx,
+				req: permissions.root,
+			},
+			id: task.id,
+			body,
+		});
+
+		// Create new task with same configuration if previous
+		// iteration completed and has a cron expression
+		if (
+			['failed', 'succeeded'].includes(body.status) &&
+			task.is_scheduled_with__cron_expression != null
+		) {
+			await this.client.post({
+				resource: 'task',
+				passthrough: {
+					tx,
+					req: permissions.root,
+				},
+				options: {
+					returnResource: false,
+				},
+				body: {
+					attempt_limit: task.attempt_limit,
+					is_created_by__actor: task.is_created_by__actor,
+					is_executed_by__handler: task.is_executed_by__handler,
+					is_executed_with__parameter_set: task.is_executed_with__parameter_set,
+					is_scheduled_with__cron_expression:
+						task.is_scheduled_with__cron_expression,
+					priority: task.priority,
+				},
+			});
+		}
+	}
+
+	// Calculate next attempt time using exponential backoff
+	private getNextAttemptTime(attempt: number): Date | null {
+		const delay = Math.ceil(Math.exp(Math.min(10, attempt)));
+		return new Date(Date.now() + delay);
+	}
+
+	// Poll for tasks to execute
+	private poll(): void {
+		let executed = false;
+		void (async () => {
+			try {
+				const handlerNames = Object.keys(this.handlers);
+				const binds = handlerNames
+					.map((_, index) => `$${index + 1}`)
+					.join(', ');
+				if (!this.canExecute()) {
+					return;
+				}
+				await sbvrUtils.db.transaction(async (tx) => {
```

Comment:
This transaction is held open whilst the tasks are being executed which could definitely cause issues if the tasks take any real amount of time, since that would mean the transaction times out/fails - can test what happens for a task that essentially does something like `await setTimeout(65000)`

Reply:
@Page- In this case the transaction is automatically closed according to `env.db.timeoutMS`, which makes sense. I want to support tasks taking a bit longer than normal transactions, but we should still have a limit. What do you think about adding an optional `options` object to `sbvrUtils.db.transaction()` to allow overriding the default `timeoutMS` that would eventually be passed to the `AutomaticClose` class? I'd like to do something generic to give task executions a bit more room without affecting other transactions.

Reply:
After talking with Page in a call, we decided to go with this `options` idea, but in a separate PR that will then be a pre-requisite for this PR. We also need to check timeouts at the Postgres level in production/staging when implementing tasks at the API level.

Reply:
@Page- Opened a PR for this here: https://github.com/balena-io/pinejs/pull/775